plaidml2/edsl/edsl_test.py

```diff
@@ -386,6 +391,213 @@ def test_use_default(self):
+            str(program), '''function (
+  I[I_0, I_1, I_2, I_3],
+  K[K_0, K_1, K_2, K_3]
+) -> (
+  _X0
+) {
+  _X0[x0, x1, x3, x6 : 1, 5, 5, 1] = +(I[x0, -1/2 + 1/2*x1 + 1/2*x2, -1/2 + 1/2*x3 + 1/2*x4, x5] * K[2 - x2, 2 - x4, x6, x5]);
+}
+''')
+        outputs = run(program, [(I, np.random.rand(1, 3, 3, 1)), (K, np.random.rand(1, 3, 3, 1))])
+
+    def testFunkyLayerNames(self):
+        '''Exercises fix for plaidml bug #241
+
+        Now that we emit keras layer names as 'pid' attribute values, in order
+        to help link tile code back to its origin while debugging, we must
+        reformat those names as valid tile identifiers. If we are doing that,
+        this test will pass, otherwise we'll get a syntax error.'''
+
+        I = Tensor(LogicalShape(plaidml.DType.FLOAT32, [3]), name='I')
+        K = Tensor(LogicalShape(plaidml.DType.FLOAT32, [3]), name='K')
+        N = TensorDim()
+        M = TensorDim()
+        i = TensorIndex()
+        j = TensorIndex()
+        I.bind_dims(N)
+        K.bind_dims(M)
+        O = TensorOutput(5)
+
+        O[i] += (I[(i - j + 1) // 2] * K[j])
+
+        program = Program('this-is-not an identifier', [O])
+        self.assertMultiLineEqual(
+            str(program), '''function (
+  I[I_0],
+  K[K_0]
+) -> (
+  _X0
+) {
+  _X0[x0 : 5] = +(I[1/2 + 1/2*x0 - 1/2*x1] * K[x1]);
+}
+''')
+        outputs = run(program, [(I, np.array([1, 2, 3])), (K, np.array([1, 2, 3]))])
+        self.assertEquals(outputs[0].tolist(), [2, 5, 4, 9, 6])
+
+    def testTileIdentity(self):
+        I = Tensor(LogicalShape(plaidml.DType.FLOAT32, [3]), name='I')
+        program = Program('tile_identity', [I])
+        self.assertMultiLineEqual(str(program), '''function (
+  I[I_0]
+) -> (
+  _X0
+) {
+  _X0 = ident(I);
+}
+''')
+        outputs = run(program, [(I, np.array([(1, 2, 3)]))])
+        self.assertEquals(outputs[0].tolist(), [1, 2, 3])
+
+    def testAssignmentExceptions(self):
+        A = Tensor(LogicalShape(plaidml.DType.FLOAT32, [5, 1]), name='A')
+        B = Tensor(LogicalShape(plaidml.DType.FLOAT32, [1, 5]), name='B')
+        L, M, N = TensorDims(3)
+        i, j, k = TensorIndexes(3)
+        A.bind_dims(L, M)
+        B.bind_dims(M, N)
+        O = TensorOutput(L, N)
+        O[i, j] = A[i, k] * B[k, j]
+        program = Program('assignment_non_exception', [O])
+        self.assertMultiLineEqual(
+            str(program), '''function (
+  A[A_0, A_1],
+  B[B_0, B_1]
+) -> (
+  _X0
+) {
+  _X0[x0, x2 : 5, 5] = =(A[x0, x1] * B[x1, x2]);
+}
+''')
+        outputs = run(program, [(A, np.array([[1], [2], [3], [4], [5]])),
+                                (B, np.array([1, 2, 3, 4, 5]))])
+        self.assertEquals(outputs[0].tolist(),
+                          [[1., 2., 3., 4., 5.], [2., 4., 6., 8., 10.], [3., 6., 9., 12., 15.],
+                           [4., 8., 12., 16., 20.], [5., 10., 15., 20., 25.]])
+
+        O = TensorOutput(L, N)
+        O[i, j] = B[j,] * A[j, k]
```

Comment:
This should be `O[i, j] = B[i, k] * A[k, j]`; and the shape of O should be `N, L` (reversed order). I would expect this test to fail (via not throwing where expected) -- if so, add an appropriate skip.

Reply:
with O[i, j] = B[i, k] * A[k, j] the test ran fine. 
See code for compared output. Not sure if this is desired behavior.

Reply:
O = TensorOutput(L, N) and O = TensorOutput(N, L) produced identical behavior

Reply:
This is not desired behavior. This test is accurately reporting that we are not throwing when an illegal assignment aggregation is requested. I recommend wrapping in a `with self.assertRaises(pml2_ffi_Error):` block and skipping the test (since it will fail).

Also: The shape I had in mind for the output was `O = TensorOutput(M, M)`, but this will also test what we want it to test with `L, N` or `N, L`.